### PR TITLE
fix(write-back): Duplicate image entries with different formats when updating two images

### PR DESCRIPTION
**Pull request by chengfang**
Fixes https://github.com/argoproj-labs/argocd-image-updater/issues/1034

Modified `mergeKustomizeOverride` method to have more control over setting and adding parameters to `.argocd-source-*.yaml` target file.

### CODE REVIEW

File: pkg/argocd/update_test.go [hunk 1]
```diff
@@ -3643,3 +3643,58 @@ func Test_GetRepositoryLock(t *testing.T) {
...
+			[]v1alpha1.KustomizeImage{"nginx:foo"},
+			[]v1alpha1.KustomizeImage{"nginx:foo"}},
+		{"with-tag-1", []v1alpha1.KustomizeImage{"nginx"},
+			[]v1alpha1.KustomizeImage{"nginx:latest"},
+			[]v1alpha1.KustomizeImage{"nginx:latest"}},
+		{"with-tag-sha", []v1alpha1.KustomizeImage{"nginx:latest"},
+			[]v1alpha1.KustomizeImage{"nginx:latest@sha256:91734281c0ebfc6f1aea979cffeed5079cfe786228a71cc6f1f46a228cde6e34"},
+			[]v1alpha1.KustomizeImage{"nginx:latest@sha256:91734281c0ebfc6f1aea979cffeed5079cfe786228a71cc6f1f46a228cde6e34"}},
+
+		{"2-images", []v1alpha1.KustomizeImage{"nginx:latest",
```
Comment (keithchong) [inline pkg/argocd/update_test.go:3664]: Cheng, do you have an example where the new image tag replaces the old image tag to test out the conditional statements in your new code in update.go?

File: pkg/argocd/update.go [hunk 2]
```diff
@@ -551,13 +551,24 @@ func mergeHelmOverride(t *helmOverride, o *helmOverride) {
...
+	for _, newImage := range *o.Kustomize.Images {
+		found := false
+		newContainerImage := image.NewFromIdentifier(string(newImage))
+		for idx, existingImage := range *t.Kustomize.Images {
+			existingContainerImage := image.NewFromIdentifier(string(existingImage))
+			if newContainerImage.ImageName == existingContainerImage.ImageName &&
+				newContainerImage.RegistryURL == existingContainerImage.RegistryURL {
+				found = true
+				if existingContainerImage.ImageTag == nil ||
+					!(existingContainerImage.ImageTag).Equals(newContainerImage.ImageTag) {
```
Comment (keithchong) [inline pkg/argocd/update.go:563]: From what I see, at least from the tests that I'm trying to run, the new image tag can be nil and so the newImage gets added

File: pkg/argocd/update.go [hunk 3]
```diff
@@ -551,13 +551,24 @@ func mergeHelmOverride(t *helmOverride, o *helmOverride) {
...
-		if idx != -1 {
-			(*t.Kustomize.Images)[idx] = image
-			continue
+	for _, newImage := range *o.Kustomize.Images {
+		found := false
+		newContainerImage := image.NewFromIdentifier(string(newImage))
+		for idx, existingImage := range *t.Kustomize.Images {
+			existingContainerImage := image.NewFromIdentifier(string(existingImage))
+			if newContainerImage.ImageName == existingContainerImage.ImageName &&
+				newContainerImage.RegistryURL == existingContainerImage.RegistryURL {
```
Comment (keithchong) [inline pkg/argocd/update.go:560]: From the tests that I performed, RegistryURL is nil all the time.  Is there another test that will test this condition with non-nil values?

File: pkg/argocd/update_test.go [hunk 4]
```diff
@@ -3643,3 +3643,58 @@ func Test_GetRepositoryLock(t *testing.T) {
 	require.NotNil(t, state.repositoryLocks[repo2])
 	require.Equal(t, lock3, state.repositoryLocks[repo2])
 }
+
+func Test_mergeKustomizeOverride(t *testing.T) {
```
Comment (keithchong) [inline pkg/argocd/update_test.go:3647]: I was trying to test the new function from a higher level, at calling `marshalParamsOverride` instead.  Do you think it is worthwhile to do it from that call?  eg. see the first couple of scenarios from the existing test `Test_MarshalParamsOverride`

File: pkg/argocd/update_test.go (hunk 1 above)
Comment (chengfang) [reply pkg/argocd/update_test.go:3664]: the sub-tests: `with-tag-1`, `with-tag-sha`, and `2-images` all replace the existing tag with the new tag.

File: pkg/argocd/update.go (hunk 2 above)
Comment (chengfang) [reply pkg/argocd/update.go:563]: In real applications, new image tag is unlikely to be nil. If it is, something already went wrong earlier.

But if the new tag is nil, it will cause nil pointer dereferece and panic. We could add a nil check to prevent that from happening:

```go
if existingContainerImage.ImageTag == nil ||
(newContainerImage.ImageTag != nil && !(existingContainerImage.ImageTag).Equals(newContainerImage.ImageTag)) {
```

File: pkg/argocd/update.go (hunk 3 above)
Comment (chengfang) [reply pkg/argocd/update.go:560]: `RegistryURL` has a default value "" and should not be nil.

The first 4 sub-tests don't use any registry url, so they should be able to test this case.

File: pkg/argocd/update_test.go (hunk 4 above)
Comment (chengfang) [reply pkg/argocd/update_test.go:3647]: I can add a sub-test to `Test_MarshalParamsOverride` by setting the new images in app spec, to test out the changes from that level.

Comment (keithchong) [reply pkg/argocd/update_test.go:3690]: Ok, I see the call to marshalParamsOverride

File: pkg/argocd/update_test.go (hunk 1 above)
Comment (keithchong) [reply pkg/argocd/update_test.go:3710]: I rechecked this again and yeah, the first test `with-tag`, the existing and the new tag are the same.  The next three tests, by definition of your test case struct, the tags are different.

**Review by keithchong — APPROVED** [review]
LGTM after re-review.

**Review by jannfis — APPROVED** [review]
LGTM
